server/aesdsocket.c

```diff
@@ -0,0 +1,303 @@
+        syslog(LOG_ERR, "ERROR set socket options failed with error number%d\n", errno);
+        shutdown_process();
+        return -1;
+    }
+
+    // Bind device address to socket
+    if (bind(server_fd, new_addr_info->ai_addr, new_addr_info->ai_addrlen) < 0) {
+        syslog(LOG_ERR, "ERROR binding socket error num %d\n", errno);
+        shutdown_process();
+        return -1;
+    }
+
+
+    // Listen for connection
+    if (listen(server_fd, MAX_NUM_CONNECTIONS)) {
+        syslog(LOG_ERR, "ERROR: listening for connection error num %d\n", errno);
+        shutdown_process();
+        return -1;
+    }
+
+    printf("Listening for connections\n\r");
+
+    if (daemonize == true) {
+        // ignore signals
+        signal(SIGCHLD, SIG_IGN);
+        signal(SIGHUP, SIG_IGN);
+        // create process
+        pid_t pid = fork();
+        if (pid < 0) {
+          shutdown_process();
+          exit(-1);
+        }
+        if (pid > 0) {
+            exit(EXIT_SUCCESS); 
+        }
+
+        // create new session
+        if (setsid() == -1) {
+            syslog(LOG_ERR, "setsid error%d\n", errno);
+            exit(-1);
+        }
+        // change dir to root dir
+        chdir("/");
+
+        //redirect to /dev/null
+        int devnull = open("/dev/null", O_RDWR);
+        dup2(devnull, STDIN_FILENO);
+        dup2(devnull, STDOUT_FILENO);
+        dup2(devnull, STDERR_FILENO);
+        close(devnull);
+      }
+
+    while(1) {
+
+        struct sockaddr_in client_addr;
+        socklen_t client_addr_size = sizeof(client_addr);
+
+        client_fd = accept(server_fd, (struct sockaddr*)&client_addr, &client_addr_size);
+        if(client_fd < 0){
+            syslog(LOG_ERR, "ERROR: accepting new connection error is %s", strerror(errno));
+            shutdown_process();
+            return -1;
+        }
+
+        char client_info[INET_ADDRSTRLEN];
+        inet_ntop(AF_INET, &(client_addr.sin_addr), client_info, INET_ADDRSTRLEN);
+
+        syslog(LOG_INFO, "Accepted connection from %s \n\r",client_info);
+        printf("Accepted connection from %s\n\r", client_info);
+        char buf[BUF_SIZE];
+        int curr_pos = 0;
+
+        while(1){
+            int read_bytes = read(client_fd, &buf[curr_pos], (BUF_SIZE - curr_pos));
+            if (read_bytes < 0) {
+                syslog(LOG_ERR, "Error: reading from socket errno=%d\n", errno);
+                curr_pos = 0; //re read
+                continue; 
+            }
+            if (read_bytes == 0)
+                break; // no bytes to read
+
+            printf("read %d bytes\n\r", read_bytes);
+            //get file opened for writing
+            int fd = open(DATA_FILE,O_RDWR | O_CREAT | O_APPEND, 0766);
+            if (fd < 0)
+                syslog(LOG_ERR, "error opening file errno is %d\n\r", errno);
+
+            lseek(fd, 0, SEEK_END); //need to append to end of file
+
+            int write_bytes = write(fd, &buf[curr_pos], read_bytes);
```

Comment:
@emma-harper what will happen if the packet is more than 1024 bytes before you receive your first '\n'? also does this comply with the requirement as stated in https://www.coursera.org/learn/advanced-embedded-software-development/supplement/N9ZTJ/assignment-5-part-1-instructions which says that the file should be written after newline? This implementation would not work for assignment 8